strawberryfields/apps/data/feature.py

```diff
@@ -0,0 +1,148 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Pre-calculated datasets of GBS kernels.
+
+.. seealso::
+
+    :doc:`/introduction/data`
+"""
+# pylint: disable=unnecessary-pass
+from abc import ABC, abstractmethod
+
+import pkg_resources
+import numpy as np
+import scipy
+
+DATA_PATH = pkg_resources.resource_filename("strawberryfields", "apps/data/FeatureData") + "/"
+
+
+class FeatureDataset(ABC):
+    """Base class for loading datasets of pre-calculated GBS feature vectors.
+
+        For each dataset, feature vectors, information on what orbits/events are used to construct
+        these feature vectors and the corresponding adjacency matrices are provided.
+
+    Attributes:
+        n_mean (float): mean number of photons used in feature vector calculation
+        threshold (bool): flag to indicate whether feature vectors are calculated with threshold
+            detection (i.e., detectors of zero or some photons) or with photon-number-resolving
+            detectors.
+        method (str): method used to calculate the feature vectors; "exact" for exact calculation
+         or "MC" for Monte Carlo estimation
+            estimation.
+        num_fv (int): number of feature vectors provided in the dataset
+        dim_fv (int): dimension of all feature vectors
+        featuresData (array): array of feature vectors
+        unitData (array): array of orbits/events used to construct the feature vectors
+        matData (array): array of adjacency matrices of graphs for which feature vectors
+            were calculated
+    """
+
+    @property
+    @abstractmethod
+    def _data_filename(self) -> str:
+        """Base name of files containing the data stored in the ``./FeatureData/`` directory.
+
+        For each dataset, feature vectors, information on what orbits/events are used to construct
+        these feature vectors and the corresponding adjacency matrices are provided in three
+        separate files.
+
+        Array of feature vectors should be provided in a text file. For ``_data_filename = "example"``,
+        feature vectors should be stored in ``./FeatureData/example_fv.txt``.
+
+        Array of corresponding adjacency matrices should be provided as a ``scipy.sparse.csr_matrix``
+        saved in ``.npz`` format. For ``_data_filename = "example"``, adjacency matrices
+        are stored in ``./FeatureData/example_mat.npz``."""

```

Comment:
ok. We can also use `numpy.savez_compressed` to stored multiple arrays in a compressed manner in a `.npz` file.

Reply:
using `np.savez` or `np.savez_compressed` has one downside: every array has to be separately called with its string name; the first graph is `npzfile['arr0']`.  when using `np.save`, we can load `.npy` file and access graph matrices as `npyfile[0]` etc.

Reply:
I had a quick look at the different options:

```python
import networkx as nx
import numpy as np
import scipy.sparse
import pickle

adjs = [nx.to_numpy_array(nx.erdos_renyi_graph(dim + 1, 0.7)) for dim in range(10)]
adjs_sp = [scipy.sparse.csc_matrix(a) for a in adjs]

# doesn't work because we have a list of sparse matrices:
# scipy.sparse.save_npz("test_sp.npz", adjs_sp)

# only works if we have allow_pickle=True which seems to be not recommended
np.save("test.npy", adjs, allow_pickle=True)
adjs2 = np.load("test.npy", allow_pickle=True)

# pickle approach seems to work but is also not recommended
with open("test.pickle", "wb") as f:
    pickle.dump(adjs, f)
    
with open("test.pickle", "rb") as f:
    adjs2 = pickle.load(f)

# this should work
np.savez("test.npz", *adjs)
loads_dict = np.load("test.npz")
adjs2 = list(loads_dict.values())
```

I think you're right, the `np.savez()` option should work! The approach above doesn't worry about giving names to the arrays, we just save them and load them in the same order (which should work since we are using Python 3.6+ in SF which has ordered dictionaries.